### PR TITLE
ethereum: implement schnorr verification contract deployment and related crypto

### PR DESCRIPTION
**Pull request by noot**
- add `schnorr-verify` contract + build script
- add crypto code in `coins/ethereum` to preprocess schnorr signature for solidity verifier
- add code in `coins/ethereum` to deploy contract and call `verify`
- add tests :D

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -1,3 +1,6 @@
 [submodule "coins/monero/c/monero"]
 	path = coins/monero/c/monero
 	url = https://github.com/monero-project/monero
+[submodule "coins/ethereum/schnorr-verify"]
+	path = coins/ethereum/schnorr-verify
+	url = https://github.com/noot/schnorr-verify.git
```
Comment (kayabaNerve) [inline .gitmodules]: Would you be open to inlining the contract from this and ensuring it has the needed tests within Rust, instead of a submodule? You would be able to maintain the copyright notice and GPL license structure :) (though I think AGPL would be better given how SCs are frequently considered backend...).

File: Cargo.toml [hunk 2]
```diff
@@ -9,6 +9,7 @@ members = [
   "crypto/dleq",
   "crypto/frost",
 
+  "coins/ethereum",
```
Comment (kayabaNerve) [inline Cargo.toml:11]: We show precedence to Ethereum over Monero? We really don't care about privacy :(

ACK. No issue here ranking by significance.

File: coins/ethereum/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+serde_json = "1.0"
+serde = "1.0"
+hex-literal = "0.3"
+ethers = { git = "https://github.com/gakonst/ethers-rs", features = ["abigen", "ethers-solc"] }
+eyre = "0.6"
+tokio = { version = "1.19", features = ["macros"] }
+k256 = { version = "0.11", features = ["arithmetic", "keccak256", "ecdsa"] }
+
+serai-processor = { path = "../../processor" }
+frost = { package = "modular-frost", path = "../../crypto/frost", features = ["secp256k1", "ed25519"] }
```
Comment (kayabaNerve) [inline coins/ethereum/Cargo.toml]: ed25519 shouldn't be a needed feature. serai-processor should import ethereum, not the other way around. I did debate if this should solely be a processor integration? Because we're not building a full code lib? But between the Schnorr support, which is generic over multisig, and the discussed inner call data retrieval... there's reason to make it a lib.

File: coins/ethereum/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,20 @@
+[package]
+name = "ethereum"
```
Comment (kayabaNerve) [inline coins/ethereum/Cargo.toml]: Mind doing ethereum-serai and filling out the other package fields, as coins/monero or the crypto libs did?

File: coins/ethereum/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+ethers = { git = "https://github.com/gakonst/ethers-rs", features = ["abigen", "ethers-solc"] }
+eyre = "0.6"
+tokio = { version = "1.19", features = ["macros"] }
+k256 = { version = "0.11", features = ["arithmetic", "keccak256", "ecdsa"] }
+
+serai-processor = { path = "../../processor" }
+frost = { package = "modular-frost", path = "../../crypto/frost", features = ["secp256k1", "ed25519"] }
+
+[dev-dependencies]
+rand = "0.8"
```
Comment (kayabaNerve) [inline coins/ethereum/Cargo.toml]: Lack of trailing new line?

File: coins/ethereum/README.md [hunk 6]
```diff
@@ -0,0 +1,23 @@
+# Ethereum
+
+This package contains Ethereum-related functionality, specifically deploying and interacting with Serai contracts.
+
+## Requirements
+
+- npm / node.js (16.0+)
```
Comment (kayabaNerve) [inline coins/ethereum/README.md]: Related to the submodule comment about moving solely to Rust.

File: coins/ethereum/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+version = "0.1.0"
+edition = "2021"
+
+[dependencies]
+thiserror = "1"
+serde_json = "1.0"
+serde = "1.0"
+hex-literal = "0.3"
+ethers = { git = "https://github.com/gakonst/ethers-rs", features = ["abigen", "ethers-solc"] }
+eyre = "0.6"
```
Comment (kayabaNerve) [inline coins/ethereum/Cargo.toml:16]: This appears to be a fork of anyhow? It'd be eligible for usage in processor, intended to be the app, yet this should just use thiserror.

File: coins/ethereum/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+tokio = { version = "1.19", features = ["macros"] }
```
Comment (kayabaNerve) [inline coins/ethereum/Cargo.toml]: Also should not be present here, except as a dev-depend. Sorry for any confusion I contributed on the processor <-> coin relation.

File: coins/ethereum/src/contract.rs [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    Schnorr,
+    "./schnorr-verify/artifacts/contracts/Schnorr.sol/Schnorr.json",
+    event_derives(serde::Deserialize, serde::Serialize),
+);
+
+pub async fn deploy_schnorr_verifier_contract(
+    client: Arc<SignerMiddleware<Provider<Http>, LocalWallet>>,
+) -> Result<schnorr_mod::Schnorr<SignerMiddleware<Provider<Http>, LocalWallet>>> {
+    let path = "./schnorr-verify/artifacts/contracts/Schnorr.sol/Schnorr.json";
+    let artifact: HardhatArtifact = serde_json::from_reader(File::open(path).unwrap()).unwrap();
```
Comment (kayabaNerve) [inline coins/ethereum/src/contract.rs]: This can be read at compile time.

Also, pub(crate) since this should be limited to tests/a `src/bin` app.

File: coins/ethereum/src/contract.rs [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+        .await?;
+    if ok {
+        Ok(())
+    } else {
+        Err(eyre!(EthereumError::VerificationError))
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (kayabaNerve) [inline coins/ethereum/src/contract.rs]: Can we please get a tests folder?

File: processor/src/coin/ethereum/mod.rs [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+            sig.as_ref()[64],
+            *sig.r(),
+            *sig.s()
+        )
+        .unwrap(),
+        address(&ProjectivePoint::from(public))
+    );
+}
+
+#[test]
```
Comment (kayabaNerve) [inline processor/src/coin/ethereum/mod.rs]: *folder*

File: processor/src/coin/ethereum/mod.rs [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+    };
+    use rand::rngs::OsRng;
+
+    let private = SigningKey::random(&mut OsRng);
+    let public = VerifyingKey::from(&private);
+
+    const MESSAGE: &'static [u8] = b"Hello, World!";
+    let sig: Signature = private.sign(MESSAGE);
+    public.verify(MESSAGE, &sig).unwrap();
+    //assert!(verify(ProjectivePoint::from(public), MESSAGE, *sig.r(), *sig.s()));
```
Comment (kayabaNerve) [inline processor/src/coin/ethereum/mod.rs]: Why is this commented?

File: processor/src/coin/ethereum/mod.rs [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+            *sig.r(),
+            *sig.s()
+        )
+        .unwrap(),
+        address(&ProjectivePoint::from(public))
+    );
+}
+
+#[test]
+fn test_signing() {
```
Comment (kayabaNerve) [inline processor/src/coin/ethereum/mod.rs]: Remove this if it isn't actually testing anything.

File: processor/src/coin/mod.rs [hunk 14]
```diff
@@ -3,82 +3,83 @@ use std::{marker::Send, sync::Arc};
...
 
+pub mod ethereum;
 pub mod monero;
 pub use self::monero::Monero;
 
 #[derive(Clone, Error, Debug)]
 pub enum CoinError {
-  #[error("failed to connect to coin daemon")]
-  ConnectionError
+    #[error("failed to connect to coin daemon")]
```
Comment (kayabaNerve) [inline processor/src/coin/mod.rs]: 2-space ident, please, not to mention the reformatting. While I'll agree to cargo fmt to resolve these debates, that should be done in a followup PR.

File: processor/src/coin/monero.rs [hunk 15]
```diff
@@ -5,220 +5,265 @@ use async_trait::async_trait;
 use curve25519_dalek::scalar::Scalar;
 
 use dalek_ff_group as dfg;
-use transcript::RecommendedTranscript;
```
Comment (kayabaNerve) [inline processor/src/coin/monero.rs]: This file as a whole should be unchanged.

File: processor/src/lib.rs [hunk 16]
```diff
@@ -1,12 +1,12 @@
...
+use std::{collections::HashMap, io::Cursor, marker::Send};
 
 use async_trait::async_trait;
 use thiserror::Error;
 
 use frost::{curve::Curve, FrostError};
 
-mod coin;
-use coin::{CoinError, Coin};
+pub mod coin;
```
Comment (kayabaNerve) [inline processor/src/lib.rs]: Shouldn't be pub. I did make comments about splitting coin out of processor once though? Offering a generic multi-coin backend for a variety of purposes? Feel free to comment there.

File: processor/src/lib.rs [hunk 17]
```diff
@@ -17,22 +17,23 @@ pub enum NetworkError {}
 
 #[async_trait]
 pub trait Network: Send {
-  async fn round(&mut self, data: Vec<u8>) -> Result<HashMap<u16, Cursor<Vec<u8>>>, NetworkError>;
```
Comment (kayabaNerve) [inline processor/src/lib.rs]: This fit in 100 chars D:

File: processor/src/lib.rs [hunk 18]
```diff
@@ -17,22 +17,23 @@ pub enum NetworkError {}
...
+    async fn round(&mut self, data: Vec<u8>)
+        -> Result<HashMap<u16, Cursor<Vec<u8>>>, NetworkError>;
 }
 
 #[derive(Clone, Error, Debug)]
 pub enum SignError {
-  #[error("FROST had an error {0}")]
```
Comment (kayabaNerve) [inline processor/src/lib.rs]: 4-space ident unrelated reformatting.

File: processor/src/tests/mod.rs [hunk 19]
```diff
@@ -1,111 +1,137 @@
-use std::{io::Cursor, sync::{Arc, RwLock}, collections::HashMap};
```
Comment (kayabaNerve) [inline processor/src/tests/mod.rs]: Same comment as the past few files.

File: processor/src/wallet.rs [hunk 20]
```diff
@@ -1,363 +1,398 @@
-use std::{sync::Arc, collections::HashMap};
```
Comment (kayabaNerve) [inline processor/src/wallet.rs]: Same comment as...

**Review by kayabaNerve — CHANGES_REQUESTED** [review]

File: .gitmodules (hunk 1 above)
Comment (noot) [reply .gitmodules]: can do, do you mean copy-pasting the entire repo basically? I'd like to eventually remove the need for JS altogether by compiling it within Rust which should be doable with ethers-solc

File: Cargo.toml (hunk 2 above)
Comment (noot) [reply Cargo.toml:11]: it's by market cap actually

File: .gitmodules (hunk 1 above)
Comment (kayabaNerve) [reply .gitmodules]: Solely the ETH contract, which would require replacing the build sys as well. So not hardhat/JS.

File: coins/ethereum/Cargo.toml (hunk 7 above)
Comment (noot) [reply coins/ethereum/Cargo.toml:16]: it's because ethers uses it, so I'd have to turn `eyre::ErrReport` into errors with `thiserror` which was a bit annoying. can update though

File: processor/src/coin/ethereum/mod.rs (hunk 12 above)
Comment (noot) [reply processor/src/coin/ethereum/mod.rs]: idk, you put this here

File: processor/src/lib.rs (hunk 16 above)
Comment (noot) [reply processor/src/lib.rs]: I think I needed to make it pub to access the ethereum functionality within `coins/ethereum`?

File: coins/ethereum/Cargo.toml (hunk 7 above)
Comment (kayabaNerve) [reply coins/ethereum/Cargo.toml:16]: Oh. That's... stupid. In that case, sure, keep it if you prefer it. I didn't want to introduce it, but if its ethers, its ethers. We may have to do the thiserror *somewhere* though, and this may be the best place to do so? Depends on how things play out... Regardless, keep for now, resolve later?

File: processor/src/coin/ethereum/mod.rs (hunk 12 above)
Comment (kayabaNerve) [reply processor/src/coin/ethereum/mod.rs]: ... does it pass? It should :thinking:

File: processor/src/coin/ethereum/mod.rs (hunk 13 above)
Comment (kayabaNerve) [reply processor/src/coin/ethereum/mod.rs]: I believe this is probably something else I put here. I think I intended for it to be used as one of the tests under coins/ethereum, lol. Or rather, that FROST-produced signatures could be verified as expected using ecrecover? So a Rust impl of the process the SC will do? Maybe this should be filled out? :thinking:

File: processor/src/lib.rs (hunk 16 above)
Comment (kayabaNerve) [reply processor/src/lib.rs]: ethers -> ethereum-serai -> processor
! ethers -> ethereum-serai <-> processor

File: processor/src/lib.rs (hunk 17 above)
Comment (kayabaNerve) [reply processor/src/lib.rs]: Seems like a Cargo fmt default. Happy we resolved to leave fmt to a follow up.

File: processor/src/coin/ethereum/mod.rs (hunk 12 above)
Comment (noot) [reply processor/src/coin/ethereum/mod.rs]: ah it's a duplicate of the above line i think, just removed it

File: processor/src/coin/ethereum/mod.rs (hunk 13 above)
Comment (noot) [reply processor/src/coin/ethereum/mod.rs]: yeah you put this here, there's already verify and ecrecover tests so we probably don't need this

File: coins/ethereum/src/contract.rs (hunk 9 above)
Comment (noot) [reply coins/ethereum/src/contract.rs]: if i move the tests to a new folder it complains that `deploy_schnorr_verifier_contract` is private if i use `pub(crate)` :/

File: coins/ethereum/Cargo.toml (hunk 3 above)
Comment (noot) [reply coins/ethereum/Cargo.toml]: updated

File: coins/ethereum/Cargo.toml (hunk 4 above)
Comment (noot) [reply coins/ethereum/Cargo.toml]: done

File: coins/ethereum/src/contract.rs (hunk 10 above)
Comment (noot) [reply coins/ethereum/src/contract.rs]: moved to folder

File: processor/src/coin/mod.rs (hunk 14 above)
Comment (noot) [reply processor/src/coin/mod.rs]: reverted all auto-fmt'd files

File: coins/ethereum/src/contract.rs (hunk 9 above)
Comment (kayabaNerve) [reply coins/ethereum/src/contract.rs]: Uhhhh `src/tests`, and then you still do `#[cfg(test)] mod tests;`

File: .gitmodules (hunk 1 above)
Comment (noot) [reply .gitmodules]: updated

File: coins/ethereum/README.md (hunk 6 above)
Comment (noot) [reply coins/ethereum/README.md]: done!

File: coins/ethereum/contracts/Schnorr.sol [hunk 21]
```diff
@@ -0,0 +1,18 @@
...
+    // er := -e*P_x
+    // px := public key x-coord
+    // parity := public key y-coord parity (27 or 28)
+    // message := 32-byte message
+    // e := schnorr signature challenge
+    function verify(bytes32 sr, bytes32 er, bytes32 px, uint8 parity, bytes32 message, bytes32 e) public view returns (bool) {
+        // ecrecover = (m, v, r, s);
+        require(sr != 0);
+        address q = ecrecover(sr, parity, px, er);
+        return e == keccak256(abi.encodePacked(q, uint8(parity), px, block.chainid, message));
```
Comment (kayabaNerve) [inline coins/ethereum/contracts/Schnorr.sol]: While I'm unsure about which 0 checks are needed and which are offered, q should equal px doesn't have to be double transcripted here.

er is not bound to e and accordingly, this is trivially forgeable.

I am curious the security of the other preprocesses, and it may be best to remove them for now. Simpler and avoids some of these issues.

File: coins/ethereum/tests/contract.rs [hunk 22]
```diff
@@ -0,0 +1,74 @@
+#[cfg(test)]
+mod tests {
```
Comment (kayabaNerve) [inline coins/ethereum/tests/contract.rs]: You don't need the mod tests if this is under tests/. You just have `mod tests;` in `lib.rs`, which is cfg test, and then `tests/mod.rs` with the tests in that file/`mod contract;` to keep tests under tests/contract.rs.

File: coins/ethereum/Cargo.toml [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+serde = "1.0"
+hex-literal = "0.3"
+
+ethers = { git = "https://github.com/gakonst/ethers-rs", features = ["abigen", "ethers-solc"] }
+eyre = "0.6"
+
+k256 = { version = "0.11", features = ["arithmetic", "keccak256", "ecdsa"] }
+frost = { package = "modular-frost", path = "../../crypto/frost", features = ["secp256k1"] }
+sha3 = "0.10"
+curve25519-dalek = { version = "3", features = ["std"] }
```
Comment (kayabaNerve) [inline coins/ethereum/Cargo.toml]: This shouldn't need dalek.

**Review by kayabaNerve — CHANGES_REQUESTED** [review]

File: coins/ethereum/Cargo.toml (hunk 23 above)
Comment (noot) [reply coins/ethereum/Cargo.toml]: removed

File: coins/ethereum/tests/contract.rs (hunk 22 above)
Comment (noot) [reply coins/ethereum/tests/contract.rs]: updated

File: coins/ethereum/contracts/Schnorr.sol (hunk 21 above)
Comment (noot) [reply coins/ethereum/contracts/Schnorr.sol]: right, added the preprocessing to the contract, it now accepts schnorr sig `(s, e)` instead. also added a comment to clarify which zero checks are needed

**Comment by kayabaNerve** [comment]
Approved, once I double check the cryptography myself :D There may still need to be an q != 0 check, as ecrecover won't panic on 0 yet will return 0. Looks solid overall though :)

**Comment by noot** [comment]
@kayabaNerve yep, I think a `q != 0` check is definitely needed, added!

**Review by kayabaNerve — APPROVED** [review]
